### ISSUE TITLE
Bookmarks: part 1

### ISSUE THREAD

**Issue by Witoso** [ckeditor/ckeditor5#17063]
### Goal

Create the first, minimal iteration of Bookmarks, so that users of the editor can create and edit them.

_This will be later followed up by a ticket which will extend the linking experience to bookmarks._

### Description

*   Anchors/Bookmarks are one of the basic features of editors. They allow placing the specific marker to navigate someone to.
*   We choose to name this feature **Bookmarks**, it’s a much more recognizable name for content creators. Its name and iconography are used in current popular content creation systems (Word, Google Docs).

### Scope

*   **Functional requirements:**
    *   CRUD for Bookmarks in the editor, with a simple inline user interface (bookmarks will be shown in the text).
    *   Bookmark creation should cover most places of the editor: text, blocks (images, tables), inline blocks, widgets.
    *   We do only point-only-bookmarks, meaning you will not be able to wrap the text/elements and place a bookmark on them. We will create it as a singular `<a id="xyz" aria-label="..."></a>`.
        *   The reasons are multiple: it’s less complex, as we don’t need to think about scenarios of a link in a link (it’s forbidden in HTML). Also, it’s less confusing for content creators, again the action of bookmarking with content that has a link leads to confusing results in the UI.
        *   That means we need to figure out where the bookmarks will appear for the block elements. For simplicity, we decided for now to place them before block elements.
    *   The UI for creation should not be confusing, especially in terms of selection. If we create point bookmarks, the creation point should be obvious.
*   **Non-functional requirements:**
    *   There shouldn’t be an observable performance penalty for bookmark creation (validation of duplicated ID).
    *   Bookmarks are accessible.
    *   Users coming with data `<a id="xyz"></a>` with GHS, should have their `a` migrated to the bookmarks UI.
    *   Uncover the tricky technical parts for further development (how pagination works, validation of names/duplicates).
    *   Should work in a collaboration session, track changes, revision history.
*   **Non-goals for this part**
    *   Handling wrapping bookmarks/anchors. `<a id="xyz">This text is in the bookmark.</a>`
        *   We need to decide in the follow-ups how we want to handle such content. Opt-in, opt-out for migration?
    *   Smaller sub-features, like copy URL/name of the Bookmark, toggling visibility of bookmarks, are not needed now. (nice to have)
    *   Interaction with other features (slash commands, document outline).
    *   Linking to bookmarks will be done in a separate part.
    *   Pasting bookmarks from Word/Gdoc.

_Prototypes_

![](https://github.com/user-attachments/assets/d11c2ee9-d6da-4eba-be09-6cce31bb08bd)

![](https://github.com/user-attachments/assets/da9f78eb-fde5-464e-ba9b-31e72e653285)

**Comment by sonfd** [comment]
Hello! This will be a great feature! Thank you for adding it. 

I'm no expert, but I have been wondering about using `<a id='my-bookmark'></a>` as the markup for a bookmark, particularly with no option to have text inside the bookmark. I stumbled on the following pages that basically recommend against that markup. I just want to post them here in case they're useful for you.

1. Stack Overflow Answer to "WCAG Anchor Link": https://stackoverflow.com/a/23749088/6335551
2. [Are your anchor links accessible?](https://amberwilson.co.uk/blog/are-your-anchor-links-accessible/) - while this article focuses on both the link and the bookmark, it seems useful nonetheless.
3. This issue in the Drupal community: [Create anchors using span tag rather than anchor tag](https://www.drupal.org/project/anchor_link/issues/3473661)

Thanks again for adding this feature!

**Comment by eckmo** [comment]
Related to the comment @sonfd left, but I'm working with a module in Drupal that does exactly this - extends CKEditor to allow for anchors. The module creates the anchor with `a` tags, which are then rendered as an `a` tag without `href`. While this is technically correct as it's semantically the same as a `span`, I believe using a `span` directly would be the better option here. 

I documented the reasoning a little better in [the Drupal module's issue queue](https://www.drupal.org/project/anchor_link/issues/3473661), but an `a` tag without an href flags in certain accessibility checkers, such as the ARC Toolkit, which may create issues in the long run. 

In my specific use case the anchor `a` tags were taking on our site's link styles as well (styles applied to all `a` tags), which was creating confusion for users as these aren't links. Adding `a:not([href])` to undo the styles certainly works, but a `span` would alleviate the need altogether.

In addition to this, the language on the [MDN page](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) regarding `a` tags makes plenty of reference to `href` in association with `a` tags, and refers to them as 'links' a lot, implying that a link is comprised of an `a` tag and its `href`, with link text indicating the link's destination. 

So TL;DR for accessibility and 'seems logically correct' reasoning, I believe a `span` is better suited for an anchor than an `a` tag. 

Stoked to have this featured directly in CKE!

**Comment by Witoso** [comment]
Thanks so much for sharing your thoughts! Firstly, the description above could have been a bit more precise. We plan to implement Bookmarks as a fully accessible `a` element, so initially we planned something like `<a id="xyz" aria-label="..."></a>`.

We also read before we started a few materials on this, also the ones that were linked above 😊 

There's always a question of how the role should be defined in the editor vs output content. In the editor, the bookmark will be interactive, so maybe some `role="button"` makes sense. Is it `presentation` in the published content? To be decided.

> a tag without an href flags in certain accessibility checkers, such as the ARC Toolkit, which may create issues in the long run.

It shouldn't be tagged, as we will provide an accessible label.

> In my specific use case the anchor a tags were taking on our site's link styles as well.

Good point! Should not happen, as bookmarks are planned to be empty elements. Of course, someone could have a global margins/paddings targeting this element, but the same goes for `span`? But fair point, targeting general `a` may be more common than general `span`.

> While this is technically correct as it's semantically the same as a span, I believe using a span directly would be the better option here.

I'm also worried about another perspective, let's call it a "recognizability". It is usually the `a` element that is used for such things throughout the systems. If you export GDoc, or copy from Word, you get the `a`'s. We often get feedback that our HTML standardization is too opinionated, so there's that too. 😅 

Curious about your thoughts, @Comandeer and @pszczesniak.

**Comment by Witoso** [comment]
Also, important to mention that this:

```html
<h2 id="introduction">Introduction</h2>
<a href="#introduction">#</a>
```

is another feature that we will plan for the future. It relates to providing auto-ids for the blocks in the editor (and auto-linking). We don't plan to focus on this part in Bookmarks to not increase the scope too much.

**Comment by niegowski** [comment]
For the editing part, we used our inline widget (the same as the inline image). While an inline widget is selected, the DOM selection is moved to the fake selection container, and the text "{bookmarkId} bookmark widget" is selected there. @Comandeer do we need to add the role and aria-label there?

![Screenshot 2024-09-20 at 10 15 38](https://github.com/user-attachments/assets/c6e5f771-20c1-482f-93c6-fbfaa20dafb6)

**Comment by Comandeer** [comment]
If the element is correctly announced by the screen reader when the selection changes (and AFAIK it is for widgets), then it's ok and we do not need to add an additional label.

**Comment by Witoso** [comment]
Work finished in this part, if anyone wants to check, the work is on `ck/epic/1944-bookmark` branch.

Next part: #17180.
